mlir/lib/Dialect/Rock/Transforms/RockPrepareLLVM.cpp

```diff
@@ -40,7 +45,148 @@ struct RockPrepareLLVMPass
+  // while we know those loads are vector-aligned (and this is similar for
+  // stores).
+  //
+  // We'd also like to reinforce that the loads we're doing from readonly
+  // arguments are invariant - concurrent modification of any input we read is
+  // undefined behavior.
+  //
+  // The second set of annotations has to do with a deficiency in the AMDGPU
+  // backend. Specifically, the `noalias` attributes on kernel arguments
+  // get discarded in the backend as the function is rewritten to include the
+  // actual kernel argument loads being performed.
+  size_t n = func.getNumArguments();
+  llvm::SmallBitVector isReadonly(n);
+  // All this alias scope stuff can be removed if the backend fixes things.
+  auto domain =
+      b.getAttr<LLVM::AliasScopeDomainAttr>(b.getStringAttr(func.getSymName()));
+  llvm::SmallVector<ArrayAttr> aliasScopes;
+  aliasScopes.reserve(n);
+  llvm::SmallVector<ArrayAttr> noaliasScopes;
+  noaliasScopes.reserve(n);
+  for (size_t i = 0; i < n; ++i) {
+    if (func.getArgAttr(i, LLVM::LLVMDialect::getReadonlyAttrName()))
+      isReadonly[i] = true;
+    if (!isa<LLVM::LLVMPointerType>(func.getArgument(i).getType())) {
+      aliasScopes.push_back(nullptr);
+      continue;
+    }
+    auto aliasScope =
+        LLVM::AliasScopeAttr::get(domain, b.getStringAttr("arg" + Twine(i)));
+    aliasScopes.push_back(b.getArrayAttr(aliasScope));
+  }
+  {
+    SmallVector<Attribute> allButOneScope;
+    allButOneScope.reserve(n);
+    for (size_t i = 0; i < n; ++i) {
+      for (auto [j, val] : llvm::enumerate(aliasScopes)) {
+        if (j != i && val)
+          allButOneScope.push_back(val[0]);
+      }
+      noaliasScopes.push_back(b.getArrayAttr(allButOneScope));
+      allButOneScope.clear();
+      LLVM_DEBUG(llvm::dbgs() << noaliasScopes.back() << "\n");
+    }
+  }
+  llvm::DenseMap<Value, BlockArgument> toArgCache;
+
+  // The alias analysis interface will pick up ROCDL buffer ops and the native
+  // LLVM ops.
+  func.walk([&](LLVM::AliasAnalysisOpInterface aliasIface) {
+    // We will make the simplyfying assumption that the first pointer-valued
+    // operand to the operation is the pointer being accessed.
+    Operation *aliasOp = aliasIface.getOperation();
+    Value ptrArg;
+    for (Value arg : aliasOp->getOperands()) {
+      if (isa<LLVM::LLVMPointerType>(arg.getType())) {
+        ptrArg = arg;
+        break;
+      }
+    }
+    if (!ptrArg)
+      return;
+    if (auto load = dyn_cast<LLVM::LoadOp>(aliasOp))
+      load.setAlignment(getAlign(load.getType()));
+    if (auto store = dyn_cast<LLVM::StoreOp>(aliasOp))
+      store.setAlignment(getAlign(store.getValue().getType()));
+    BlockArgument funcArg = traceToArg(ptrArg, func, toArgCache);
+    if (!funcArg)
+      return;
+    unsigned argNo = funcArg.getArgNumber();
+    if (auto load = dyn_cast<LLVM::LoadOp>(aliasOp))
+      load.setInvariantLoad(isReadonly[argNo]);
```

Comment:
Not sure how much is relevant, but I stumbled upon this: https://github.com/rust-lang/rust/pull/103070 , which talks about `invariant.start`. Do you mind making sure that the compilation time is not heavily affected by this?